articles/gosagawa.re

```diff
@@ -0,0 +1,61 @@
+= Goのエラーハンドリングを深掘りする
+
+== 概要
+
+アプリケーションを実装するにあたり、エラーハンドリングは非常に重要です。しかし、エラー管理がデフォルトでは非常にシンプルです。カスタマイズをしなければ不便であるところは欠点でありつつもの、逆にしっかりとカスタマイズをすれば非常に強力なツールとなりえます。この記事ではpkg/errorsライブラリや、GopherCon2019のセッション"Handling go errors",そしてAthensプロジェクトでのエラーハンドリングの
+実例を通して、実用にみあったエラーハンドリング処理の書き方をお伝えできればと思っています。
+
+== エラーハンドリングで満たしたい要件
+
+Goのエラーハンドリングの良いところでもあり大変なところでもあるのは、カスタマイズがしやすく設計が必要であるという事だと思います。
+何もせず、そのままreturn errを繰り返すだけだとエラーのテキスト情報のみしか受け渡す事ができません。
+
+では、実際の業務でエラーハンドリングに求められる要件はどのような事があるでしょうか？
+
+Webアプリケーションを例に取った時、以下がよくある要件なのではと思います。
+
+* エラーを階層構造でもち、オリジナルのエラーと最後にラップしたエラー含めていずれの階層のエラーも取得できる
+* 問題の発生箇所のファイル、関数名、行数を確認できる
+* エラーの種類、ログレベル、エラーコード、レスポンスコードを保持できる
+* アプリケーション固有の情報を保持できる
+
+一つ一つみていきましょう。
+* エラーを階層構造でもち、オリジナルのエラーと最後にラップしたエラー含めていずれの階層のエラーも取得できる
+
+簡単なプログラムでない限り、エラーは階層化されていくかと思います。例えば、ログイン処理でユーザ取得時にDB接続に失敗した場合
+DB接続エラー、ユーザ取得エラー、ログインエラーと階層化されます。こうなった時にログインエラーだけわかっても何が原因かわかりませんし、
+DB接続エラーだけみてもそれがどのような影響をるのかわからないため必要になるかと思い明日。
+
+* 問題の発生箇所のファイル、行数を確認できる
+エラーが発生したら、発生箇所のコードを追う事になるので欲しい情報です。なお、階層化された場合階層化されたエラーそれぞれの発生箇所がわかるのが望ましいです。
+
+* エラーの種類、ログレベル、エラーコード、レスポンスコードを保持できる
```

Comment:
**[textlint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
This sentence exceeds the maximum count of comma. Maximum is 3. (ja-technical-writing/max-comma)